In [1]:
!pwd

/Users/akhichoudhary/STATS/Thinkstats/Thinkstats2_exercise/Thinkstats/Approach any problem with ml


In [6]:
import os

file_path = '/Users/akhichoudhary/STATS/Thinkstats/Thinkstats2_exercise/Thinkstats/Approach any problem with ml/car-in-the-dat.zip'
print(os.path.exists(file_path))  # Should return True


False


In [24]:
# import zipfile
# import pandas as pd

    # df=pd.read_csv('/Users/akhichoudhary/STATS/Thinkstats/Thinkstats2_exercise/Thinkstats/Approach any problem with ml/data/train.csv')

# or-------

# # Path to the ZIP folder
# zip_path = '/Users/akhichoudhary/STATS/Thinkstats/Thinkstats2_exercise/Thinkstats/Approach any problem with ml/cat-in-the-dat.zip'

# # Open the zip and list contents
# with zipfile.ZipFile(zip_path, 'r') as z:
#     print("Files in ZIP:", z.namelist())  # View files like 'train.csv', 'test.csv'

#     # Load specific CSVs
#     with z.open('train.csv') as f:
#         train_df = pd.read_csv(f)

#     with z.open('test.csv') as f:
#         test_df = pd.read_csv(f)

# # Now you can use train_df and test_df as usual
# # print(train_df.head())
# # print(test_df.head())


## Start

In [17]:
mapping={
    "Freezing": 0,
    "Warm": 1,
    "Cold": 2,
    "Boiling Hot": 3,
    "Hot": 4,
    "Lava Hot": 5
    
}

In [18]:
train_df.ord_2.value_counts()

ord_2
Freezing       99816
Lava Hot       63908
Boiling Hot    60627
Cold           33768
Hot            22227
Warm           19654
Name: count, dtype: int64

In [19]:
train_df.loc[:,"ord_2"]=train_df.ord_2.map(mapping)

In [20]:
train_df.ord_2.value_counts()

ord_2
0    99816
5    63908
3    60627
2    33768
4    22227
1    19654
Name: count, dtype: int64

## ~ LabelEncoder,OneHotEncoder,Binarization
- above same encoding into numerical value can be done by LabelEncoder from scikit-learn

### Label Encoder

In [22]:
from sklearn import preprocessing
df=pd.read_csv('/Users/akhichoudhary/STATS/Thinkstats/Thinkstats2_exercise/Thinkstats/Approach any problem with ml/data/train.csv')

In [25]:
df.loc[:,"ord_2"]=df.ord_2.fillna("NONE")
# LabelEncoder from scikit-learn does not handle NaN values, 

In [26]:
lbl_enc=preprocessing.LabelEncoder()

In [27]:
# fit label encoder and transform values on ord_2 column
# P.S: do not use this directly. fit first, then transform
df.loc[:,"ord_2"]=lbl_enc.fit_transform(df.ord_2.values)

-  checking the bytes used by different techniques

In [29]:
import numpy as np
example=np.array([
    [0,0,1],
    [1,0,0],
    [1,0,1]]
)
print(example.nbytes)

72


In [30]:
from scipy import sparse

example=np.array(
    [[0, 0, 1],
[1, 0, 0],
[1, 0, 1]]
)

sparse_example=sparse.csr_matrix(example)
print(sparse_example.data.nbytes)

32


- The total size of the sparse csr matrix is the sum of three values.

In [32]:
print(
    sparse_example.data.nbytes +
sparse_example.indptr.nbytes +
sparse_example.indices.nbytes
)

64


In [33]:
# number of rows
n_rows = 10000
# number of columns
n_cols = 100000
# create random binary matrix with only 5% values as 1s
example = np.random.binomial(1, p=0.05, size=(n_rows, n_cols))
# print size in bytes
print(f"Size of dense array: {example.nbytes}")
# convert numpy array to sparse CSR matrix
sparse_example = sparse.csr_matrix(example)
# print size of this sparse matrix
print(f"Size of sparse array: {sparse_example.data.nbytes}")
full_size = (
sparse_example.data.nbytes +
sparse_example.indptr.nbytes +
sparse_example.indices.nbytes
)
# print full size of this sparse matrix
print(f"Full size of sparse array: {full_size}")

Size of dense array: 8000000000
Size of sparse array: 399997896
Full size of sparse array: 600036848


In [35]:
# So, dense array takes ~8000MB or approximately 8GB of memory. The sparse
# array, on the other hand, takes only 399MB of memory.
# And, that’s why we prefer sparse arrays over dense whenever we have a lot of zeros
# in our features.

### Binarization

In [36]:
import numpy as np
from scipy import sparse
# create binary matrix
example = np.array(
[
[0, 0, 0, 0, 1, 0],
[0, 1, 0, 0, 0, 0],
[1, 0, 0, 0, 0, 0]
]
)
# print size in bytes
print(f"Size of dense array: {example.nbytes}")
# convert numpy array to sparse CSR matrix
sparse_example = sparse.csr_matrix(example)
# print size of this sparse matrix
print(f"Size of sparse array: {sparse_example.data.nbytes}")
full_size = (
sparse_example.data.nbytes +
sparse_example.indptr.nbytes +
sparse_example.indices.nbytes
)
# print full size of this sparse matrix
print(f"Full size of sparse array: {full_size}")

Size of dense array: 144
Size of sparse array: 24
Full size of sparse array: 52


### OneHotEncoding

In [39]:
import numpy as np
from sklearn import preprocessing
# create random 1-d array with 1001 different categories (int)
example = np.random.randint(1000, size=1000000)
# initialize OneHotEncoder from scikit-learn
# keep sparse = False to get dense array
ohe = preprocessing.OneHotEncoder(sparse_output=False)
# fit and transform data with dense one hot encoder
ohe_example = ohe.fit_transform(example.reshape(-1, 1))
# print size in bytes for dense array
print(f"Size of dense array: {ohe_example.nbytes}")
# initialize OneHotEncoder from scikit-learn
# keep sparse = True to get sparse array
ohe = preprocessing.OneHotEncoder(sparse_output=True)
# fit and transform data with sparse one-hot encoder
ohe_example = ohe.fit_transform(example.reshape(-1, 1))
# print size of this sparse matrix
print(f"Size of sparse array: {ohe_example.data.nbytes}")
full_size = (
ohe_example.data.nbytes +
ohe_example.indptr.nbytes + ohe_example.indices.nbytes
)
# print full size of this sparse matrix
print(f"Full size of sparse array: {full_size}")

Size of dense array: 8000000000
Size of sparse array: 8000000
Full size of sparse array: 16000004


### Data manipulation

In [40]:
df.groupby(["ord_2"])["id"].transform("count")


0         33768
1         22227
2         63908
3         60627
4         99816
          ...  
299995    99816
299996    99816
299997    60627
299998    60627
299999    99816
Name: id, Length: 300000, dtype: int64

In [41]:
df.groupby(['ord_1','ord_2'])['id'].count().reset_index(name='count')

ord_1  ord_2  count
0   Contributor      0   8692
1   Contributor      1   4842
2   Contributor      2  14284
3   Contributor      3   3122
4   Contributor      4   9074
5   Contributor      5   2857
6        Expert      0   4980
7        Expert      1   2850
8        Expert      2   8432
9        Expert      3   1887
10       Expert      4   5274
11       Expert      5   1642
12  Grandmaster      0  15719
13  Grandmaster      1   8702
14  Grandmaster      2  25620
15  Grandmaster      3   5697
16  Grandmaster      4  16617
17  Grandmaster      5   5073
18       Master      0   5720
19       Master      1   3129
20       Master      2   9401
21       Master      3   2069
22       Master      4   5882
23       Master      5   1852
24       Novice      0  25516
25       Novice      1  14245
26       Novice      2  42079
27       Novice      3   9452
28       Novice      4  27061
29       Novice      5   8230

- is used to count the number of rows (based on id) for each combination of ord_1 and ord_2 values in your DataFrame.
- .reset_index(name="count")
Resets the index (so you get a flat DataFrame) and renames the id count column to "count".

In [43]:
# create new feature
df['new feature']=df.ord_1.astype(str)+"_"+df.ord_2.astype(str)
df['new feature'].head(5)

0    Grandmaster_1
1    Grandmaster_3
2         Expert_4
3    Grandmaster_0
4    Grandmaster_2
Name: new feature, dtype: object

#### Steps
Whenever you get categorical variables, follow these simple steps:
- • fill the NaN values (this is very important!)
- • convert them to integers by applying label encoding using LabelEncoder of scikit-learn or by using a mapping dictionary. If you didn’t fill up NaN
values with something, you might have to take care of them in this step
- • create one-hot encoding. Yes, you can skip binarization!


#### "rare" Category
- A rare category is a category
which is not seen very often and can include many different categories. You can
also try to “predict” the unknown category by using a nearest neighbour model
- if you predict this category, it will become one of the categories from
the training data.
- we can build a simple model
that’s trained on all features except “f3”(feature column where rare value is there). Thus, you will be creating a model that
predicts “f3” when it’s not known or not available in training.

🧠 If you design your cross-validation in such a way that it
replicates the prediction process when you run your model on test data, then it’s
never going to overfit.

for e.g 💡 Suppose you want to concatenate training and test data, then in each
fold you must concatenate training and validation data and also make sure that your
validation dataset replicates the test set. 

In [48]:
import pandas as pd
from sklearn import preprocessing
# read training data
train =pd.read_csv('/Users/akhichoudhary/STATS/Thinkstats/Thinkstats2_exercise/Thinkstats/Approach any problem with ml/data/train.csv')
#read test data
test =pd.read_csv('/Users/akhichoudhary/STATS/Thinkstats/Thinkstats2_exercise/Thinkstats/Approach any problem with ml/data/test.csv')
# create a fake target column for test data
# since this column doesn't exist
test.loc[:, "target"] = -1
# concatenate both training and test data
data = pd.concat([train, test]).reset_index(drop=True)
# make a list of features we are interested in
# id and target is something we should not encode
features = [x for x in train.columns if x not in ["id", "target"]]
# loop over the features list
for feat in features:
    # create a new instance of LabelEncoder for each feature
    lbl_enc = preprocessing.LabelEncoder()
    # note the trick here
    # since its categorical data, we fillna with a string
    # and we convert all the data to string type
    # so, no matter its int or float, its converted to string
    # int/float but categorical!!!
    temp_col = data[feat].fillna("NONE").astype(str).values
    # we can use fit_transform here as we do not
    # have any extra test data that we need to
    # transform on separately
    data.loc[:, feat] = lbl_enc.fit_transform(temp_col)
    # split the training and test data again
train = data[data.target != -1].reset_index(drop=True)
test = data[data.target == -1].reset_index(drop=True)

In [55]:
# df =pd.read_csv('/Users/akhichoudhary/STATS/Thinkstats/Thinkstats2_exercise/Thinkstats/Approach any problem with ml/data/train.csv')

df.ord_2.fillna("NONE").value_counts()

ord_2
Freezing       99816
Lava Hot       63908
Boiling Hot    60627
Cold           33768
Hot            22227
Warm           19654
Name: count, dtype: int64

- 🧠💡 We can treat “NONE” as unknown. So, if during live testing, we get new categories
that we have not seen before, we will mark them as “NONE”.

In [58]:
len(df)

300000

In [60]:
df.ord_4.fillna("NONE").value_counts()

ord_4
L    19066
G    18899
S    18875
A    18258
R    16927
Q    15925
K    14698
I    14644
Z    14232
T    14220
V    14143
J    12878
P    12839
U    12775
H    12743
F    11717
E    11303
W     9197
Y     8490
X     6292
B     6169
O     5836
D     3974
C     3575
N     2166
M      159
Name: count, dtype: int64

- We see that some values appear only a couple thousand times, and some appear
almost 40000 times.
- define our criteria for calling a value “rare”. Let’s say the requirement
for a value being rare in this column is a count of less than 3000
- and all missing values will be mapped to
“NONE”.(in data set not preset but in other case)

In [68]:
df.loc[df.ord_4.value_counts()[df.ord_4].values<3000,'ord_4']="RARE"

In [69]:
df.ord_4.value_counts()

ord_4
L       19066
G       18899
S       18875
A       18258
R       16927
Q       15925
K       14698
I       14644
Z       14232
T       14220
V       14143
J       12878
P       12839
U       12775
H       12743
F       11717
E       11303
W        9197
Y        8490
X        6292
B        6169
O        5836
D        3974
C        3575
RARE     2325
Name: count, dtype: int64

### Building our model

Before going to any kind of model building, it’s essential to take care of cross-
validation. We have already seen the label/target distribution, and we know that it
is a binary classification problem with skewed targets. Thus, we will be using
StratifiedKFold to split the data here

#### Creat_folds for cat-in-data

/Users/akhichoudhary/STATS/Thinkstats/Thinkstats2_exercise/Thinkstats/Approach any problem with ml/mnist_classifier/src/ohe_logres.py:17: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '['0' '1' '2' ... '299997' '299998' '299999']' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  df.loc[:, col] = df[col].astype(str).fillna("NONE")
/Users/akhichoudhary/STATS/Thinkstats/Thinkstats2_exercise/Thinkstats/Approach any problem with ml/mnist_classifier/src/ohe_logres.py:17: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '['0' '0' '0' ... '0' '0' '0']' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  df.loc[:, col] = df[col].astype(str).fillna("NONE")
/Users/akhichoudhary/STATS/Thinkstats/Thinkstats2_exercise/Thinkstats/Approach any problem with ml/mnist_classifier/src/ohe_logres.py:17: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '['0' '0' '0' ... '0' '0' '1']' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  df.loc[:, col] = df[col].astype(str).fillna("NONE")
/Users/akhichoudhary/STATS/Thinkstats/Thinkstats2_exercise/Thinkstats/Approach any problem with ml/mnist_classifier/src/ohe_logres.py:17: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '['0' '0' '1' ... '0' '0' '0']' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  df.loc[:, col] = df[col].astype(str).fillna("NONE")
/Users/akhichoudhary/STATS/Thinkstats/Thinkstats2_exercise/Thinkstats/Approach any problem with ml/mnist_classifier/src/ohe_logres.py:17: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '['1' '1' '3' ... '1' '1' '3']' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  df.loc[:, col] = df[col].astype(str).fillna("NONE")
/Users/akhichoudhary/STATS/Thinkstats/Thinkstats2_exercise/Thinkstats/Approach any problem with ml/mnist_classifier/src/ohe_logres.py:17: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '['3' '1' '4' ... '1' '2' '5']' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  df.loc[:, col] = df[col].astype(str).fillna("NONE")
/Users/akhichoudhary/STATS/Thinkstats/Thinkstats2_exercise/Thinkstats/Approach any problem with ml/mnist_classifier/src/ohe_logres.py:17: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '['3' '3' '4' ... '2' '3' '5']' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  df.loc[:, col] = df[col].astype(str).fillna("NONE")
/Users/akhichoudhary/miniconda3/envs/ml/lib/python3.10/site-packages/sklearn/linear_model/_logistic.py:470: ConvergenceWarning: lbfgs failed to converge after 100 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=100).
You might also want to scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
-- $$ 0.7916089250181116 $$

In [73]:
!pwd

/Users/akhichoudhary/STATS/Thinkstats/Thinkstats2_exercise/Thinkstats/Approach any problem with ml


In [ ]:
# python -W ignore src/ohe_logres.py
Fold = 0, AUC = 0.7916089250181116
Fold = 1, AUC = 0.7885835002477297
Fold = 2, AUC = 0.7908998540691035
Fold = 3, AUC = 0.7898881988857341
Fold = 4, AUC = 0.7915466071107451